compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -0,0 +1,886 @@
+
+  /** Check usage of terms inside types
+   *
+   *  This is intended to avoid type soundness issues in Dotty.
+   */
+  def checkTermUsage(tpt: Tree, thisV: Value, klass: ClassSymbol): Contextual[List[Error]] =
+    val buf = new mutable.ArrayBuffer[Error]
+    val traverser = new TypeTraverser {
+      def traverse(tp: Type): Unit = tp match {
+        case TermRef(_: SingletonType, _) =>
+          buf ++= cases(tp, thisV, klass, tpt).errors
+        case _ =>
+          traverseChildren(tp)
+      }
+    }
+    traverser.traverse(tpt.tpe)
+    buf.toList
+
+}
+
+object Semantic {
+
+// ----- Utility methods and extractors --------------------------------
+
+  def typeRefOf(tp: Type)(using Context): TypeRef = tp.dealias.typeConstructor match {
+    case tref: TypeRef => tref
+    case hklambda: HKTypeLambda => typeRefOf(hklambda.resType)
+  }
+
+  opaque type Arg  = Tree | ByNameArg
+  case class ByNameArg(tree: Tree)
+
+  extension (arg: Arg)
+    def isByName = arg.isInstanceOf[ByNameArg]
+    def tree: Tree = arg match
+      case t: Tree      => t
+      case ByNameArg(t) => t
+
+  object Call {
+
+    def unapply(tree: Tree)(using Context): Option[(Tree, List[List[Arg]])] =
+      tree match
+      case Apply(fn, args) =>
+        val argTps = fn.tpe.widen match
+          case mt: MethodType => mt.paramInfos
+        val normArgs: List[Arg] = args.zip(argTps).map {
+          case (arg, _: ExprType) => ByNameArg(arg)
+          case (arg, _)           => arg
+        }
+        unapply(fn) match
+        case Some((ref, args0)) => Some((ref, args0 :+ normArgs))
+        case None => None
+
+      case TypeApply(fn, targs) =>
+        unapply(fn)
+
+      case ref: RefTree if ref.tpe.widenSingleton.isInstanceOf[MethodicType] =>
+        Some((ref, Nil))
+
+      case _ => None
+  }
+
+  object NewExpr {
+    def unapply(tree: Tree)(using Context): Option[(TypeRef, New, Symbol, List[List[Arg]])] =
+      tree match
+      case Call(fn @ Select(newTree: New, init), argss) if init == nme.CONSTRUCTOR =>
+        val tref = typeRefOf(newTree.tpe)
+        Some((tref, newTree, fn.symbol, argss))
+      case _ => None
+  }
+
+  object PolyFun {
+    def unapply(tree: Tree)(using Context): Option[Tree] =
+      tree match
+      case Block((cdef: TypeDef) :: Nil, Typed(NewExpr(tref, _, _, _), _))
+      if tref.symbol.isAnonymousClass && tref <:< defn.PolyFunctionType
+      =>
+        val body = cdef.rhs.asInstanceOf[Template].body
+        val apply = body.head.asInstanceOf[DefDef]
+        Some(apply.rhs)
+      case _ =>
+        None
+  }
+
+  extension (symbol: Symbol) def hasSource(using Context): Boolean =
+    !symbol.defTree.isEmpty
+
+  def resolve(cls: ClassSymbol, sym: Symbol)(using Context): Symbol =
+    if (sym.isEffectivelyFinal || sym.isConstructor) sym
+    else sym.matchingMember(cls.appliedRef)
+
+  def resolveSuper(cls: ClassSymbol, superType: Type, sym: Symbol)(using Context): Symbol = {
+    import annotation.tailrec
```

Comment:
Is there something like this elsewhere in Dotty? Should there be?

Reply:
There isn't anything similar, so I think it's better to keep it here -- also because they are two important helper methods in the logic of the analysis.